### PR TITLE
Adding RLE support, fix #45 ?

### LINKED ISSUE #45: Add .rle lifeform parser to lifeforms/wiki.py
Let's add one of the most popular formats used in [LifeWiki](https://www.conwaylife.com/wiki/Main_Page)!

Format description: https://www.conwaylife.com/wiki/Run_Length_Encoded

### PR DESCRIPTION
**Pull request by eugenpt**
Added RLE parser to lifeforms/wiki.py, +some refactoring

### CODE REVIEW

**Comment by eugenpt** [comment]
Sorry, it seems I was not successful in updating my fork (since so many commits are listed in this PR..).
Or maybe the issue is that I've started to work on RLE in a separate branch before previous PR was approved and merged

**Comment by ljvmiranda921** [comment]
Sure it's fine! Sorry won't be able to check this until Sunday (or the next weekend, not this week)! Will be on vacation leave for a bit to recharge! Will find a timeto read through this and add my comments!

**Comment by eugenpt** [comment]
Have a nice vacation !

File: seagull/lifeforms/wiki.py [hunk 1]
```diff
@@ -120,25 +122,70 @@ def _get_metadata(lines: List[str]) -> Dict:
...
+    Returns:
+        str: [description]
+    """
+    if isfile(path):
+        logger.trace(f"reading from file [{path}]..", end="")
+        with open(path, "r") as f:
+            content = f.read()
+        logger.trace("ok")
+    elif urlparse(path).scheme in {"ftp", "http", "https"}:
+        # web-hosted file?
```
Comment (ljvmiranda921) [inline seagull/lifeforms/wiki.py:174]:
```suggestion
```

File: seagull/lifeforms/wiki.py [hunk 2]
```diff
@@ -120,25 +122,70 @@ def _get_metadata(lines: List[str]) -> Dict:
...
+        logger.trace(f"trying to download [{path}]..", end="")
+        req = urlopen(path)
+        if req.getcode() != 200:
+            raise ValueError(
+                "Invalid input URL" f" request returned {req.getcode()}"
```
Comment (ljvmiranda921) [inline seagull/lifeforms/wiki.py:179]: You can put this as one string:

```suggestion
                f"Invalid input URL: request returned {req.getcode()}"
```

File: seagull/lifeforms/wiki.py [hunk 3]
```diff
@@ -209,3 +240,157 @@ def parse_cells(cells_str: str) -> Lifeform:
...
+    Does not add "!" at the end, converts only commands
+        (idea behind this is that it insures that you know what you're doing)
+
+    Parameters
+    ----------
+    cells_str : Union[list, str]
+        Plaintext format lifeform description
+        may be a list of lines or a str, possibly multiline
+
+
```
Comment (ljvmiranda921) [inline seagull/lifeforms/wiki.py:257]:
```suggestion
```

File: seagull/lifeforms/wiki.py [hunk 4]
```diff
@@ -209,3 +240,157 @@ def parse_cells(cells_str: str) -> Lifeform:
...
+    Returns
+    -------
+    str
+        in RLE format, no header, no comments, no trailing "!"
+
```
Comment (ljvmiranda921) [inline seagull/lifeforms/wiki.py:262]:
```suggestion
```

File: seagull/lifeforms/wiki.py [hunk 5]
```diff
@@ -209,3 +240,157 @@ def parse_cells(cells_str: str) -> Lifeform:
...
+    """
+    if isinstance(cells_str, str):
+        cells_str = cells_str.replace("\r\n", "\n").split("\n")
+
+    cells_str = "\n".join(l for l in cells_str if not l.startswith("!"))
+
```
Comment (ljvmiranda921) [inline seagull/lifeforms/wiki.py:268]:
```suggestion
```

File: seagull/lifeforms/wiki.py [hunk 6]
```diff
@@ -209,3 +240,157 @@ def parse_cells(cells_str: str) -> Lifeform:
...
+    blocks = re.findall("(\n+|\.+|O+)", cells_str)
+
```
Comment (ljvmiranda921) [inline seagull/lifeforms/wiki.py:270]:
```suggestion
```

File: seagull/lifeforms/wiki.py [hunk 7]
```diff
@@ -209,3 +240,157 @@ def parse_cells(cells_str: str) -> Lifeform:
...
+    parse_dict = {"\n": "$", ".": "b", "O": "o"}
+
```
Comment (ljvmiranda921) [inline seagull/lifeforms/wiki.py:272]:
```suggestion
```

File: seagull/lifeforms/wiki.py [hunk 8]
```diff
@@ -209,3 +240,157 @@ def parse_cells(cells_str: str) -> Lifeform:
...
+    )
+    if header_match is None:
+        raise ValueError(
+            f"Incorrect input: wrong header line format : [{header_line}]"
+        )
+    header_match = header_match.groups()
+
+    # Hm.. do these really matter if I do not parse #R,#P,#r comments?
+    #  possible usage: check dimension of parser plaintext layout
+    #   which could differ really if rle is not optimized
```
Comment (ljvmiranda921) [inline seagull/lifeforms/wiki.py:376]: I like what you did above where you put `@TODO` and `@FIXME` for things you're not sure or can still be optimized. I suggest doing the same thing here :+1:

**Comment by ljvmiranda921** [comment]
Let me check the failing builds, mostly a dependencies thing!

**Comment by ljvmiranda921** [comment]
This is now merged via #47 ! Thanks so much for this @eugenpt !
If you are interested, you can add documentation via Jupyter Notebooks like in [here](https://github.com/ljvmiranda921/seagull/tree/master/docs/notebooks). This will then show up in our [official documentation](https://pyseagull.readthedocs.io/en/latest/examples.html)!

**Pull request closed without merging**